### DISCUSSION TITLE
Is bundling the NodeJS binary necessary?

### DISCUSSION THREAD

**Discussion by jwetzell** [bitfocus/companion#2591] [Q&A] [answered]
### Is this a feature relevant to companion itself, and not a module?

- [X] I believe this to be a feature for companion, not a module

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the feature

I am in the process of creating an application that is closely modeled after the layout of the companion app (electron launcher that spawns a NodeJS process). I initially borrowed a lot of the bundling node executable logic from this application. Then I found out it is possible to spawn the child process using the node runtime that is being included with the electron launcher. I did a very brief test removing the node "bundling" portion of the [package.mjs](https://github.com/bitfocus/companion/blob/master/tools/build/package.mjs) script and altering the launcher main.js and everything appeared to run as expected.

I can think of some immediate concerns that might arise by simply using the node runtime provided by electron but was curious if this has been explored before.

**Comment by Julusian** [comment, marked as answer]
We used to just use electron and fork off it when needed, but that meant that we were always lagging behind on electron version as updating electron would risk breaking things especially when they changed major nodejs version.

We have had problems in the past when updating the nodejs version, because it would break some of the community written modules. This separate nodejs binary is largey preparation so that we can bundle multiple versions, and use different ones for different modules.

Then theres that time that:
1) libjpeg-turbo broke because electron/chromium has a version builtin, and the version we were loading conflicted with it
2) [electron broke](https://www.electronjs.org/blog/v8-memory-cage) many node-api modules, and can have lower performance as a consequence (when working with buffers from native modules)
3) I'm sure there are more times where things completely broke but I cant think of them right now.
Side note: I have spent way too much time fighting electron or electron-builder over the years, I am getting fed up with electron and its ecosystem

Long term, I want to look at replacing electron, as it is a rather heavy solution for the 'tiny' launcher that we have, but that wasnt a motivation here.

A nice bonus for us, is that it is very easy to use the same builds on a headless linux server. You simply ignore the electron binary, and launch the nodejs inner directly https://github.com/bitfocus/companion-pi/blob/main/launch.sh#L15

But for your project, using a separate nodejs is most likely unnecessary. The benefit for us is mostly due to the hundreds of community maintained modules and their potentially limited nodejs version support.

**Reply by jwetzell** [reply]
That was pretty much the exact answer I expected! I figured that the vast amount of modules would complicate things for companion's use case.

As I said I modeled things very (very) closely to the companion layout and thus the actual meat of the app (main.js) is just a commander based cli that the launcher calls (simply `npx showbridge` or by docker image is my route for "headless"). I too would love a launcher solution that isn't as overkill as electron.

Thanks for the quick reply! and boy does your name come up in just about project I mess around with (show control, network protocol libs, device control, etc.).

I'd love to dig a little more into the companion core and give back some day!
